Create 3.1 migration guide (#362)
saleor/saleor-docs: docs/developer/extending/apps/extending-dashboard-with-apps.mdx

---
title: Extending Dashboard with Apps
---

## Key concepts

App extensions allow extending different dashboard views by mounting the extension app inside a popup or redirecting users to the main app's page.
The app is created as a secure iframe that protects against XSS attacks.
It provides a way for customizing and extending the dashboard while keeping the code separate.

## Manifest

> Apps can have multiple extensions. The extension is added by specifying it in the app's manifest.
> The below sample manifest defines two extensions that will be attached to the dashboard:

Reviewer: ```suggestion
A single App can provide multiple extensions. Each extension is added by specifying it in the app's manifest.
The example manifest below defines two extensions, one providing a custom product action that opens a modal overlay, and a second one providing an alternative product creation page:
```

```json
{
  ...
  "extensions": [
      {
        "label": "Create with Sample app",
        "mount": "PRODUCT_DETAILS_MORE_ACTIONS",
        "target": "POPUP",
        "permissions": [
            "MANAGE_PRODUCTS"
        ],
        "url": "https://example.com/extension/"
      },
      {
        "label": "Create with App and redirect",
        "mount": "PRODUCT_OVERVIEW_CREATE",
        "target": "APP_PAGE",
        "permissions": [
            "MANAGE_PRODUCTS"
        ],
        "url": "/extension/redirect"
      }
    ],
}
```

> - `label` - Name which will be displayed in the dashboard
> - `mount` - Place, where extension will be mounted
> - `target` - Way of opening an extension's url. Default to `POPUP`.
>   `POPUP` will mount extension in an iframe view. `APP_PAGE` will redirect users to the dashboard's app page
> - `url` - Extension url. Url can start with `/` when `target` is set to `APP_PAGE` or manifest has defined `appUrl`.

Reviewer: ```suggestion
- `label`: name which will be displayed in the dashboard
- `mount`: the place where the extension will be mounted
- `target`: method of presenting the interface (defaults to `POPUP`);
  `POPUP` will present the interface in a modal overlay, `APP_PAGE` will navigate to the application page
- `url`: URL of the view to display; you can skip the domain and protocol when `target` is set to `APP_PAGE`, or when your manifest defines an `appUrl`.
```

>   When `target` is set to `POPUP`, `url` will be used to render an iframe

Reviewer: ```suggestion
  When `target` is set to `POPUP`, the `url` will be used to render an `<iframe>`.
```

## Possible mounting places

> Saleor requires that extension will define a place of mounting. The below table explains all mounting places available in Saleor.

Reviewer: ```suggestion
Saleor requires extensions to define a mounting place. The table below explains all mounting locations currently supported by Saleor.
```
